mountpoint-s3/src/prefetch/parquet_stream.rs

```diff
@@ -71,32 +70,54 @@ where
             let span = debug_span!("prefetch", range=?range);
             let key = key.to_owned();
 
+            let ref_cache = self
+                .cache
+                .entry(ObjectId::new(key.to_owned(), if_match.clone()))
+                .or_default();
+
+            let cached_entry = ref_cache.value().clone();
+
             async move {
-                let is_parquet_file = key.ends_with(".parquet");
-                let metadata_available = raw_metadata.read().await.is_some();
+                // Check if metadata is already stored in the cache, if not, read, parse and store the metadata
+                if key.ends_with(".parquet") && !cached_entry.state.read().await.is_some() {
+                    let mut state = cached_entry.state.write().await;
+                    if state.is_none() {
+                        let result = parse_parquet_metadata(
+                            client.clone(),
+                            &bucket,
+                            &key,
+                            &if_match,
+                            range.object_size() as u64,
+                        )
+                        .await;
 
-                if is_parquet_file && metadata_available {
+                        if let Ok(cache_value) = result {
```

Comment:
What seems to be missing here is handling the error case. Do we need to add a third state where we tried to load/parse the parquet metadata and failed? Otherwise we'll keep trying at every new request.